### PR TITLE
Point clouds 3d classification rendering

### PR DESCRIPTION
**Pull request by NEDJIMAbelgacem**
## Description
Following what @nyalldawson did in https://github.com/qgis/QGIS/pull/40412
Classifications rendering with visibility selection

classification

For now, in the opengl shader I am using discard instruction in the fragment shader which makes performance slower. I tried setting the alpha value in the produced 1D texture to 0 but it didn't work.

### CODE REVIEW

**Comment by NEDJIMAbelgacem** [comment]
Update:
> For now, in the opengl shader I am using discard instruction in the fragment shader which makes performance slower. I tried setting the alpha value in the produced 1D texture to 0 but it didn't work.

Fixed by not passing the filtered out categories to the vertex buffer at all

File: src/3d/symbols/qgspointcloud3dsymbol_p.cpp [hunk 1]
```diff
@@ -521,4 +521,118 @@ Qt3DRender::QGeometry *QgsRGBPointCloud3DSymbolHandler::makeGeometry( Qt3DCore::
...
+    if ( attrIsX )
+      iParam = x;
+    else if ( attrIsY )
+      iParam = y;
+    else if ( attrIsZ )
+      iParam = z;
+    else
+      context.getAttribute( ptr, i * recordSize + attributeOffset, attributeType, iParam );
+
+    if ( context.isFilteredOut( ( int ) iParam ) )
```
Comment (nyalldawson) [inline src/3d/symbols/qgspointcloud3dsymbol_p.cpp:621]: There's potentially a substantial speed boost you'd get if you instead built an unordered set of the filtered values just outside of this loop (ie around line 598) and tested values here using the set instead. It would certainly be much easier for the compiler to optimise vs the un-inlineable call to isFilteredOut.

File: src/3d/shaders/pointcloud.frag [hunk 2]
```diff
@@ -16,14 +16,14 @@ uniform sampler1D u_colorRampTexture; //
 // Sets the color ramp value count, used to check the if not void
 uniform int u_colorRampCount;
 
-vec3 linearColorRamp()
+vec4 linearColorRamp()
```
Comment (nyalldawson) [inline src/3d/shaders/pointcloud.frag:19]: is this change still required?

File: src/3d/symbols/qgspointcloud3dsymbol.cpp [hunk 3]
```diff
@@ -367,3 +367,147 @@ void QgsRgbPointCloud3DSymbol::setBlueContrastEnhancement( QgsContrastEnhancemen
...
+QgsColorRampShader QgsClassificationPointCloud3DSymbol::colorRampShader() const
+{
+  QgsColorRampShader colorRampShader;
+  colorRampShader.setColorRampType( QgsColorRampShader::Type::Exact );
+  colorRampShader.setClassificationMode( QgsColorRampShader::ClassificationMode::Continuous );
+  QList<QgsColorRampShader::ColorRampItem> colorRampItemList;
+  for ( const QgsPointCloudCategory &category : mCategoriesList )
+  {
+    QColor color = category.color();
+    color.setAlphaF( category.renderState() ? 1.0f : 0.0f );
```
Comment (nyalldawson) [inline src/3d/symbols/qgspointcloud3dsymbol.cpp:478]: Can we drop this now and just skip those unwanted categories?

File: src/app/3d/qgspointcloud3dsymbolwidget.h [hunk 4]
```diff
@@ -60,6 +61,7 @@ class QgsPointCloud3DSymbolWidget : public QWidget, private Ui::QgsPointCloud3DS
   private:
     int mBlockChangedSignals = 0;
     int mDisableMinMaxWidgetRefresh = 0;
+    QgsPointCloudClassifiedRendererWidget *mClassifiedRenderer = nullptr;
```
Comment (nyalldawson) [inline src/app/3d/qgspointcloud3dsymbolwidget.h:64]:
```suggestion
    QgsPointCloudClassifiedRendererWidget *mClassifiedRendererWidget = nullptr;
```

**Review by nyalldawson — COMMENTED** [review]
Looking great so far!

**Comment by nyalldawson** [comment]
I've given this branch a test run and it's working nicely!

**Comment by nyalldawson** [comment]
@NEDJIMAbelgacem you'll also need to apply a similar fix to https://github.com/qgis/QGIS/pull/40477 to this branch

File: src/3d/shaders/pointcloud.frag (hunk 2 above)
Comment (NEDJIMAbelgacem) [reply src/3d/shaders/pointcloud.frag:19]: Returning RGBA values directly looks better to me

File: src/3d/symbols/qgspointcloud3dsymbol.cpp (hunk 3 above)
Comment (NEDJIMAbelgacem) [reply src/3d/symbols/qgspointcloud3dsymbol.cpp:478]: Yes, we don't need that line anymore

**Comment by nyalldawson** [comment]
LGTM!

File: src/3d/shaders/pointcloud.frag [hunk 5]
```diff
@@ -81,33 +81,31 @@ vec3 exactColorRamp()
     vec3 color=colorRampLine.yzw;
     float value=colorRampLine.x;
     if ( abs( parameter - value ) < 0.01 )
-      return color;
+      return vec4( color, 1.0f );
   }
-
-  return vec3(0.5, 0.5, 0.5);
+  return vec4(0.0, 0.0, 0.0, 0.0f);
```
Comment (wonder-sk) [inline src/3d/shaders/pointcloud.frag:86]: This is a bit confusing to return a color with zero alpha if we're not having alpha blending enabled... Maybe let's keep alpha=1 to avoid us getting confused why it's like that at some point later :-)

File: src/3d/symbols/qgspointcloud3dsymbol_p.cpp [hunk 6]
```diff
@@ -521,4 +521,118 @@ Qt3DRender::QGeometry *QgsRGBPointCloud3DSymbolHandler::makeGeometry( Qt3DCore::
...
+bool QgsClassificationPointCloud3DSymbolHandler::prepare( const QgsPointCloud3DRenderContext &context )
+{
+  Q_UNUSED( context )
+  return true;
+}
+
+void QgsClassificationPointCloud3DSymbolHandler::processNode( QgsPointCloudIndex *pc, const IndexedPointCloudNode &n, const QgsPointCloud3DRenderContext &context )
+{
+  QgsPointCloudAttributeCollection attributes;
+  constexpr int xOffset = 0;
```
Comment (wonder-sk) [inline src/3d/symbols/qgspointcloud3dsymbol_p.cpp:539]: Did you mean 'const' here? That's what is used for yOffset and zOffset too - good to keep this consistent although there might not be much difference between constexpr and const here...

**Review by wonder-sk — COMMENTED** [review]
Looking good to me as well

File: src/3d/symbols/qgspointcloud3dsymbol_p.cpp (hunk 6 above)
Comment (NEDJIMAbelgacem) [reply src/3d/symbols/qgspointcloud3dsymbol_p.cpp:539]: I just copied whatever was in QgsColorRampPointCloud3DSymbolHandler. It seems to be one of the changes @nyalldawson did some days ago. I will change constexpr to const since yOffset and zOffset variables are const
